Develop/kwargs

mbourqui opened the pull request:
Update download() signature to use keyword arguments instead of positional arguments. For example, the downloader can now be called like
```
import ncbi_genome_download as ngd
ngd.download(group='bacteria', assembly_level='complete')
```
Argument parser still works and uses same default values as method signature. Default values are stored in a customized Enum to ensure immutability and provide some ease of use.

Tests are also updated.

kblin commented on ncbi_genome_download/__init__.py:
```diff
@@ -7,7 +7,7 @@
 )
 
-__version__ = '0.2.3'
+__version__ = '0.2.4'
```
Please don't bump the version number in an unrelated commit.

mbourqui replied:
Fixed in 5fb4e40.

kblin commented on ncbi_genome_download/core.py:
```diff
@@ -1,74 +1,193 @@
     from requests.packages.urllib3.contrib import pyopenssl
-    pyopenssl.inject_into_urllib3()
 
+    pyopenssl.inject_into_urllib3()
```
I'd prefer to leave the inject close to the import, as that keeps the "hack old pythons to properly support ssl" block.

mbourqui replied:
Done by the auto code reformatting feature of my IDE. Fixed it.

kblin commented on ncbi_genome_download/core.py:
```diff
@@ -1,74 +1,193 @@
-def download(args):
-    '''Download data from NCBI'''
+@unique
+class EMap(Enum):
```
Nice use of an Enum here. I'm not used to a bunch of the new Python 3 features yet, but this does make the code easier to read.

mbourqui replied:
Thank you!

kblin commented on ncbi_genome_download/core.py:
```diff
@@ -174,7 +190,11 @@ def download(**kwargs):
-    # TODO: warning/error if unrecognized option
+    # FIXME: improve error handling and feedback
+    assert len(kwargs) == 0, "Unrecognized option(s)"
+    if len(kwargs) > 0:
```
That code is unreachable, as the line above already exits if `len(kwargs) != 0`.

mbourqui replied:
Dropped unreachable code.

kblin commented on tests/test_core.py:
```diff
@@ -19,175 +20,176 @@ def req():
+    download_mock = mocker.MagicMock()
+    monkeypatch.setattr(core, 'download', download_mock)
+    core.download(domain='all', output='/tmp/fake')
+    assert download_mock.call_count == len(core.SUPPORTED_DOMAINS)
```
Does this one still pass? It seems that in the current code, the `download` function is only called once.

mbourqui replied:
According to http://github.drone.secondarymetabolites.org/kblin/ncbi-genome-download/44, this test passes.
`download` is called once, but `_download` is called once for each taxonomic group.

mbourqui replied:
Indeed, in that code version, the test was not passing. It is now passing again, after being updated to the changes of `core.py`.

kblin commented on tests/test_core.py:
```diff
@@ -369,22 +378,28 @@ def test_grab_checksums_file(req):
-'''
+"""
     expected = [
-        {'checksum': 'd3c2634cedd0efe05cbf8a5f5384d921', 'file': 'GCF_000009605.1_ASM960v1_feature_table.txt.gz'},
```
I know editors don't like long lines, but in this case I feel this change hurts the readability, rather than helping.

mbourqui replied:
Was done automatically by my IDE. Fixed it.

kblin commented on ncbi_genome_download/__main__.py:
```diff
@@ -9,21 +9,21 @@
-                        choices=dflt.DOMAIN.choices,
-                        default=dflt.DOMAIN.default,
-                        help='The NCBI "domain" to download (default: %(default)s)')
+    parser.add_argument('group',
```
NCBI calls this "domain", so I'd like to stick to that term, even if it's wrong.

mbourqui replied:
I found in that document ftp://ftp.ncbi.nlm.nih.gov/pub/factsheets/HowTo_Downloading_Genomic_Data.pdf (in the *Genomes FTP Directory Structure* section) that they were refering to it as taxonomic group, thus my naming change.
But I can revert to domain if you'd like to.

kblin replied:
I can't find the README file I had for this. I guess the NCBI gets to rename their stuff if they like. For sure "taxonomic group" is a more correct term anyway. Let's just keep it for now.

mbourqui commented:
The first time I tried the tests locally, I had some path issues. So I moved the test folder in 2e2f7e3f. But I moved it back to its initial location in 49d6cd8 as the tests were then running fine again.

kblin commented:
Ok, playing with this a bit more, it seems that now the command line tool doesn't work anymore. I should have a fix for that, though.

mbourqui commented:
Try with my latest commit

kblin commented on ncbi_genome_download/__main__.py:
```diff
@@ -2,9 +2,9 @@
-from . import EDefaults as dflt
+from ncbi_genome_download import __version__
+from ncbi_genome_download import download
+from ncbi_genome_download import EDefaults as dflt
```
Where are you seeing this as a problem? That bit wasn't the issue for me.
Rather `download(vars(args))` clearly needs to be `download(**vars(args))`, and then you also need to remove the "debug", "verbose", and "retries" entry.

mbourqui replied:
I encountered locally the issue I fixed. Maybe it just affects Python 3.
Working on the other issues.

kblin replied:
Possibly. I started testing on 2.7 :)

mbourqui commented:
Should be fine now!

kblin commented:
It also looks like the default file format changed from genbank to cds-fasta.

kblin commented:
That's on python 2.7 only, though.

kblin commented:
Yeah, cPython does not guarantee that the order of dictionaries is preserved. It's an implementation detail that this happens in cPython 3.6.

mbourqui commented:
Ouch, that's an issue. I was using Enums because of the ordering garantee...

kblin commented:
Well, looking at the enum34 code, it seems to be using an OrderedDict, so I assume there might be an issue in `Emap.keys()`

kblin commented on ncbi_genome_download/core.py:
```diff
@@ -97,6 +97,8 @@ def get_content(cls, key):
 
 class EFormats(EMap):
+    __order__ = 'GENBANK FASTA FEATURES GFF PROTFASTA GENREPT WGS CDSFASTA RNAFASTA'
+    # only needed in Python 2
```
Maybe put the comment above the line if it doesn't fit at the end?

mbourqui replied:
Certainly!

kblin commented:
Yup, this works, thanks a bunch.

kblin commented:
Thanks for the contribution, and your fast response.

mbourqui commented:
You're welcome, thank you for your base code! :)
